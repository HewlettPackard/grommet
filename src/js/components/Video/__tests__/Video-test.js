import React from 'react';
import renderer from 'react-test-renderer';
import 'jest-styled-components';
import { cleanup, render, fireEvent } from '@testing-library/react';
import { Grommet, Video } from '../..';

const CONTENTS = [
  <source key="source" src="small.mp4" type="video/mp4" />,
  <track key="track" />,
];
describe('Video', () => {
  afterEach(cleanup);

  test('renders', () => {
    const component = renderer.create(
      <Grommet>
        <Video>{CONTENTS}</Video>
      </Grommet>,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('autoPlay renders', () => {
    const component = renderer.create(
      <Grommet>
        <Video autoPlay>{CONTENTS}</Video>
      </Grommet>,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('loop renders', () => {
    const component = renderer.create(
      <Grommet>
        <Video loop>{CONTENTS}</Video>
      </Grommet>,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('mute renders', () => {
    const component = renderer.create(
      <Grommet>
        <Video mute>{CONTENTS}</Video>
      </Grommet>,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('controls renders', () => {
    const component = renderer.create(
      <Grommet>
        <Video controls="over">{CONTENTS}</Video>
        <Video controls="below">{CONTENTS}</Video>
      </Grommet>,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('fit renders', () => {
    const component = renderer.create(
      <Grommet>
        <Video fit="cover">{CONTENTS}</Video>
        <Video fit="contain">{CONTENTS}</Video>
      </Grommet>,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('Play and Pause event handlers', () => {
    const { container } = render(
      <Grommet>
        <Video controls="below" playing={false}>
          {CONTENTS}
        </Video>
      </Grommet>,
    );
    const videoContainer = document.querySelector('video');
    fireEvent.play(videoContainer);
    expect(container.firstChild).toMatchSnapshot();
    fireEvent.pause(videoContainer);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('End event handler', () => {
    const { container } = render(
      <Grommet>
        <Video controls="below">{CONTENTS}</Video>
      </Grommet>,
    );
    // Need to fire play event to get video playing before we fire ended event.
    const videoContainer = document.querySelector('video');
    fireEvent.play(videoContainer);
    fireEvent.ended(videoContainer);
    expect(container.firstChild).toMatchSnapshot();
  });

  test(' Configure Menu Button', () => {
    window.scrollTo = jest.fn();
    const { container, getByLabelText } = render(
      <Grommet>
        <Video controls="below" playing={false}>
          {CONTENTS}
        </Video>
      </Grommet>,
    );
    fireEvent.click(getByLabelText('Open Menu'));
    expect(container.firstChild).toMatchSnapshot();
  });

  test('fullscreen button', () => {
    window.scrollTo = jest.fn();
    console.warn = jest.fn();
    const warnSpy = jest.spyOn(console, 'warn');
    const { getByLabelText } = render(
      <Grommet>
        <Video controls="below" playing={false}>
          {CONTENTS}
        </Video>
      </Grommet>,
    );
    fireEvent.click(getByLabelText('Open Menu'));
    fireEvent.click(getByLabelText('Expand'));
    /* expect warn to have been called because jest doesn't test in any browser,
    will always have warning here due to the jest browser not supporting
     fullscreen */
    expect(warnSpy).toHaveBeenCalledWith(
      "This browser doesn't support fullscreen.",
    );
    warnSpy.mockReset();
    warnSpy.mockRestore();
    console.warn.mockRestore();
    window.scrollTo.mockRestore();
  });

  test('play button', () => {
    const playStub = jest
      .spyOn(window.HTMLMediaElement.prototype, 'play')
      .mockImplementation(() => {});
    const { getByLabelText } = render(
      <Grommet>
        <Video controls="below" playing={false}>
          {CONTENTS}
        </Video>
      </Grommet>,
    );
    fireEvent.click(getByLabelText('Play'));
    expect(playStub).toHaveBeenCalled();
    playStub.mockRestore();
  });

  test('volume controls', () => {
    const volMock = jest.fn();
    window.scrollTo = jest.fn();
    const { getByLabelText } = render(
      <Grommet>
        <Video controls="below" playing={false} onVolumeChange={volMock}>
          {CONTENTS}
        </Video>
      </Grommet>,
    );
    fireEvent.click(getByLabelText('Open Menu'));
    fireEvent.click(getByLabelText('VolumeLow'));
    expect(volMock).toHaveBeenCalled();

    window.scrollTo.mockRestore();
  });
});
