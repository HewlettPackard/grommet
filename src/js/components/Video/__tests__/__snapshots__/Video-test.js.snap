// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Video autoplay renders 1`] = `
.c0 {
  font-family: 'Work Sans',Arial,sans-serif;
  font-size: 1em;
  line-height: 1.5;
  color: #333333;
  background-color: #FFFFFF;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

.c0 * {
  box-sizing: inherit;
}

.c1 {
  -webkit-flex: 1 1;
  -ms-flex: 1 1;
  flex: 1 1;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  overflow: hidden;
  position: relative;
}

.c2 {
  max-width: 100%;
}

.c2::cue {
  background: rgba(0,0,0,0.7);
}

@media only screen and (min-width:481px) {
  .c0 {
    top: 0px;
    bottom: 0px;
    left: 0px;
    right: 0px;
    height: 100%;
    width: 100%;
    overflow: visible;
  }
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    <video
      autoPlay={false}
      className="c2"
      loop={false}
      onAbort={[Function]}
      onCanPlay={[Function]}
      onCanPlayThrough={[Function]}
      onDurationChange={[Function]}
      onEmptied={[Function]}
      onEncrypted={[Function]}
      onEnded={[Function]}
      onError={[Function]}
      onLoadStart={[Function]}
      onLoadedData={[Function]}
      onLoadedMetadata={[Function]}
      onPause={[Function]}
      onPlay={[Function]}
      onPlaying={[Function]}
      onProgress={[Function]}
      onRateChange={[Function]}
      onSeeked={[Function]}
      onSeeking={[Function]}
      onStalled={[Function]}
      onSuspend={[Function]}
      onTimeUpdate={[Function]}
      onVolumeChange={[Function]}
      onWaiting={[Function]}
    >
      <source />
      <track />
    </video>
  </div>
</div>
`;

exports[`Video renders 1`] = `
.c0 {
  font-family: 'Work Sans',Arial,sans-serif;
  font-size: 1em;
  line-height: 1.5;
  color: #333333;
  background-color: #FFFFFF;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

.c0 * {
  box-sizing: inherit;
}

.c1 {
  -webkit-flex: 1 1;
  -ms-flex: 1 1;
  flex: 1 1;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  overflow: hidden;
  position: relative;
}

.c2 {
  max-width: 100%;
}

.c2::cue {
  background: rgba(0,0,0,0.7);
}

@media only screen and (min-width:481px) {
  .c0 {
    top: 0px;
    bottom: 0px;
    left: 0px;
    right: 0px;
    height: 100%;
    width: 100%;
    overflow: visible;
  }
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    <video
      autoPlay={false}
      className="c2"
      loop={false}
      onAbort={[Function]}
      onCanPlay={[Function]}
      onCanPlayThrough={[Function]}
      onDurationChange={[Function]}
      onEmptied={[Function]}
      onEncrypted={[Function]}
      onEnded={[Function]}
      onError={[Function]}
      onLoadStart={[Function]}
      onLoadedData={[Function]}
      onLoadedMetadata={[Function]}
      onPause={[Function]}
      onPlay={[Function]}
      onPlaying={[Function]}
      onProgress={[Function]}
      onRateChange={[Function]}
      onSeeked={[Function]}
      onSeeking={[Function]}
      onStalled={[Function]}
      onSuspend={[Function]}
      onTimeUpdate={[Function]}
      onVolumeChange={[Function]}
      onWaiting={[Function]}
    >
      <source />
      <track />
    </video>
  </div>
</div>
`;
